LUSDChickenBonds/src/BLUSDLPZap.sol

```diff
@@ -0,0 +1,99 @@
+    address constant BLUSD_LUSD_3CRV_LP_TOKEN_ADDRESS = 0x5ca0313D44551e32e0d7a298EC024321c4BC59B4;
+    address constant BLUSD_LUSD_3CRV_GAUGE_ADDRESS= 0xdA0DD1798BE66E17d5aB1Dc476302b56689C2DB4;
+
+    uint256 constant LUSD_3CRV_POOL_FEE_DENOMINATOR = 10 ** 10;
+
+    IERC20 constant public lusdToken = IERC20(LUSD_TOKEN_ADDRESS);
+    IERC20 constant public bLUSDToken = IERC20(BLUSD_TOKEN_ADDRESS);
+    ICurvePool constant public lusd3CRVPool = ICurvePool(LUSD_3CRV_POOL_ADDRESS);
+    ICurveCryptoPool constant public bLUSDLUSD3CRVPool = ICurveCryptoPool(BLUSD_LUSD_3CRV_POOL_ADDRESS);
+    IERC20 constant public bLUSDLUSD3CRVLPToken = IERC20(BLUSD_LUSD_3CRV_LP_TOKEN_ADDRESS);
+    ICurveLiquidityGaugeV5 constant public bLUSDGauge = ICurveLiquidityGaugeV5(BLUSD_LUSD_3CRV_GAUGE_ADDRESS);
+
+    // TODO: add permit version
+    function _addLiquidity(
+        uint256 _bLUSDAmount,
+        uint256 _lusdAmount,
+        uint256 _minLPTokens,
+        address _receiver
+    )
+        internal
+        returns (uint256 bLUSDLUSD3CRVTokens)
+    {
+        require(_bLUSDAmount > 0 || _lusdAmount > 0, "BLUSDLPZap: Cannot provide zero liquidity");
+
+        uint256 lusd3CRVAmount;
+        if (_lusdAmount > 0) {
+            lusdToken.transferFrom(msg.sender, address(this), _lusdAmount);
+
+            // add LUSD single sided
+            lusdToken.approve(address(lusd3CRVPool), _lusdAmount);
+            lusd3CRVAmount = lusd3CRVPool.add_liquidity([_lusdAmount, 0], 0, address(this));
+        }
+
+        bLUSDToken.transferFrom(msg.sender, address(this), _bLUSDAmount);
+
+        // add bLUSD/LUSD-3CRV
+        bLUSDToken.approve(address(bLUSDLUSD3CRVPool), _bLUSDAmount);
+        if (lusd3CRVAmount > 0) {
+            lusd3CRVPool.approve(address(bLUSDLUSD3CRVPool), lusd3CRVAmount);
+        }
+        bLUSDLUSD3CRVTokens = bLUSDLUSD3CRVPool.add_liquidity([_bLUSDAmount, lusd3CRVAmount], _minLPTokens, false, _receiver);
+
+        return bLUSDLUSD3CRVTokens;
+    }
+
+    function addLiquidity(uint256 _bLUSDAmount, uint256 _lusdAmount, uint256 _minLPTokens) external returns (uint256 bLUSDLUSD3CRVTokens) {
+        // add liquidity
+        bLUSDLUSD3CRVTokens = _addLiquidity(_bLUSDAmount, _lusdAmount, _minLPTokens, msg.sender);
+
+        return bLUSDLUSD3CRVTokens;
+    }
+
+    function addLiquidityAndStake(
+        uint256 _bLUSDAmount,
+        uint256 _lusdAmount,
+        uint256 _minLPTokens
+    )
+        external
+        returns (uint256 bLUSDLUSD3CRVTokens)
+    {
+        // add liquidity
+        bLUSDLUSD3CRVTokens = _addLiquidity(_bLUSDAmount, _lusdAmount, _minLPTokens, address(this));
+
+        // approve LP tokens to Gauge
+        bLUSDLUSD3CRVLPToken.approve(address(bLUSDGauge), bLUSDLUSD3CRVTokens);
+
+        // stake into gauge
+        bLUSDGauge.deposit(bLUSDLUSD3CRVTokens, msg.sender, false); // make sure rewards are not claimed
```

Comment:
OK so here we deposit on behalf of the user, and I assume rewards accrue to their address (and they can claim later)

Reply:
Yes, that’s my understanding reading the code:

https://etherscan.io/address/0xdA0DD1798BE66E17d5aB1Dc476302b56689C2DB4#code

![Screenshot_20221011_124725](https://user-images.githubusercontent.com/701095/195070860-b4e66581-0b37-4a15-ac9d-2c35454d9205.png)